backend/cfg/cfg_dataflow.ml

```diff
@@ -132,3 +132,192 @@ module Forward (D : Domain) (T : Transfer with type domain = D.t) :
     done;
     if !iteration < max_iteration then Result.Ok res else Result.Error res
 end
+
+(* CR xclerc for xclerc: unify with `Domain` or rename. *)
+module type DomainXXX = sig
+  type t
+
+  val bot : t
+
+  val compare : t -> t -> int
+
+  val join : t -> t -> t
+
+  val less_equal : t -> t -> bool
+
+  val to_string : t -> string
+end
+
+(* CR xclerc for xclerc: unify with `Transfer` or rename. *)
+module type TransferXXX = sig
+  type domain
+
+  val basic : domain -> exn:domain -> Cfg.basic Cfg.instruction -> domain
+
+  val terminator :
+    domain -> exn:domain -> Cfg.terminator Cfg.instruction -> domain
+
+  val exception_ : domain -> domain
+end
+
+(* CR xclerc for xclerc: unify with `S` or rename. *)
+module type SXXX = sig
+  type domain
+
+  type map = domain Label.Tbl.t
+
+  val run :
+    Cfg.t -> ?max_iteration:int -> init:domain -> unit -> (map, map) Result.t
+end
+
+module Backward (D : DomainXXX) (T : TransferXXX with type domain = D.t) :
+  SXXX with type domain = D.t = struct
+  (* CR xclerc for xclerc: see what can be shared with `Forward`. *)
+
+  type domain = D.t
+
+  type map = domain Label.Tbl.t
+
+  module WorkSetElement = struct
+    type t =
+      { label : Label.t;
+        value : domain
+      }
+
+    let compare { label = left_label; value = left_value }
+        { label = right_label; value = right_value } =
+      match Label.compare left_label right_label with
+      | 0 -> D.compare left_value right_value
+      | res -> res
+  end
+
+  module WorkSet = Set.Make (WorkSetElement)
+
+  let transfer_block : domain -> exn:domain -> Cfg.basic_block -> domain =
+   fun value ~exn block ->
+    ListLabels.fold_right block.body
+      ~init:(T.terminator value ~exn block.terminator) ~f:(fun instr value ->
+        T.basic value ~exn instr)
+
+  let is_exit_block : Cfg.basic_block -> bool =
+   fun block ->
+    match block.terminator.desc with
+    | Never -> assert false
+    | Always _ -> false
+    | Parity_test _ -> false
+    | Truth_test _ -> false
+    | Float_test _ -> false
+    | Int_test _ -> false
+    | Switch _ -> false
+    | Return -> true
+    | Raise _ -> block.trap_depth = 1
+    | Tailcall _ -> true
+    | Call_no_return _ -> true
+
+  let create : Cfg.t -> init:domain -> map * WorkSet.t ref =
+   fun cfg ~init ->
+    let map = Label.Tbl.create (Label.Tbl.length cfg.Cfg.blocks) in
+    let set = ref WorkSet.empty in
+    let value = init in
+    Cfg.iter_blocks cfg ~f:(fun label block ->
+        if is_exit_block block || true
```

Comment:
Just to follow up on an offline discussion with @xclerc @stedolan about this fixpoint computation:  either all nodes go into the worklist initially, or we need to have a special value "unvisited" that is distinct from the bottom of the domain.